### PR TITLE
DOC: Quickstart Guide updates

### PR DESCRIPTION
**Pull request by mdhaber**
Updating quickstart guides, including adding instructions for Windows Subsystem for Linux. This will give Windows users a relatively convenient way to develop SciPy locally. 

*Update: Docker Quickstart (gh-10687) is now the one recommended for Windows users.*

### CODE REVIEW

File: doc/source/dev/contributor/quickstart_ubuntu.rst [hunk 1]
```diff
@@ -14,7 +14,8 @@ This quickstart guide will cover:
 * performing an in-place build of SciPy; and
 * creating a virtual environment that adds this development version of SciPy to the Python path
 
-in Ubuntu.
+in Ubuntu 16.04. *Users running Windows can follow these
+instructions after setting up* `Windows Subsystem for Linux`_ *or an Amazon EC2 instance with Ubuntu 16.04 LTS.*
```
Comment (tylerjereddy) [inline doc/source/dev/contributor/quickstart_ubuntu.rst]: perhaps slightly orthogonal to the discussion, but not completely irrelevant since I've done some releases using WSL--I can't actually tell if the newer "more native" versions of WSL allow:

1. Use of Docker on Ubuntu on Windows -- IIRC this is surprisingly complex because the low-level things are still slightly different from pure Linux.
2. Use of apps that open X11-style windows on Ubuntu on Windows

The version I'm using is probably one of the earlier ones, where neither of those is possible.
Or more accurately, you have to look for blog posts and go through involved work-arounds rather than having native support for that. Maybe not true anymore though!

File: doc/source/dev/contributor/quickstart_ubuntu.rst [hunk 2]
```diff
@@ -31,7 +32,9 @@ Building SciPy
...
+ `In a terminal, navigate <https://help.ubuntu.com/community/UsingTheTerminal>`_ to the location in which you'd like to install Anaconda. You can download the file using the terminal command ``curl -O URL_OF_FILE``, where ``URL_OF_FILE`` is to be replaced by the URL of the Anaconda installer ``.sh`` file found at the Anaconda Distribution website. When the download is complete, enter ``chmod +x file.sh``, where ``file.sh`` is to be replaced with the full name of the downloaded file. This ensures that the file's permissions allow it to be executed. Next, enter ``./file.sh``, where ``file.sh`` is again to be replaced with the full name of the downloaded file. This starts the installation process. From there, simply follow the prompts, including the "Next Steps" at the end after the installer finishes.
+
+#. `Rename the file`_ ``anaconda3/lib/libgfortran.so`` to ``anaconda3/lib/libgfortran.so_backup``, where ``anaconda3`` is to be replaced with the full path of your Anaconda installation. This file provides an incorrect Fortran shared library; renaming it forces the system to find the right one.
```
Comment (tylerjereddy) [inline doc/source/dev/contributor/quickstart_ubuntu.rst]: I definitely recognize this as something that sometimes has to be done. A little unfortunate--I would have thought / hoped there was a way to install the older runtime libraries needed through conda-forge or something like that.

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 1 above)
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: I can't speak to #1. For #2, I was able to do [this](https://www.howtogeek.com/261575/how-to-run-graphical-linux-desktop-applications-from-windows-10s-bash-shell/). Unfortunately, I couldn't use Spyder via X11 forwarding. (In case anyone is interested - an alternative to that is to run Spyder on Windows and have it execute code on a kernel running on WSL. I've had a little trouble editing files on WSL in the Spyder running on Windows, but it sounds like it's [possible, only the run button won't work](https://github.com/spyder-ide/spyder/issues/3065).)

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 2 above)
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: *Sometimes* is right. It's been tough to find a set of instructions that will work the same way every time. 

By the time I left Sunday, I had a development version of SciPy working in WSL Ubuntu 18.04 (compiled using Conda's gcc_linux-64, etc...) and in WSL Ubuntu 16.04 (compiled with LinuxBrew gcc). I updated Windows, and afterwards neither worked. I don't remember what the problems were, but I tried a bunch of things and nothing seemed to work, so I decided to start from scratch. 

I couldn't get SciPy to compile in WSL Ubuntu 16.04 with conda's compilers. I ran into the `collect2: error: ld returned 1 exit status` mentioned in gh-8325, but the fixes suggested there didn't work for me.

I was somewhat successful with compilers installed with `apt`. The tests run with one failure from within python (`import scipy; scipy.test()`) but won't run with `runtests.py`; If I remember correctly I get the `undefined symbol: _gfortran_stop_numeric_f08` error during collection.

I'll work on this more after CZI is done.

**Comment by endolith** [comment]
> This will give Windows users a relatively convenient way to develop SciPy locally.

Could also just have a downloadable virtual machine image that would Just Work on any OS?  I'm thinking VirtualBox .ova, but maybe Windows Subsystem for Linux has a similar "appliance" concept?

File: doc/source/dev/contributor/quickstart_ubuntu.rst [hunk 3]
```diff
@@ -31,7 +32,9 @@ Building SciPy
...
 
-   In Ubuntu, the Anaconda download comes as a ``.sh`` file. `In a terminal, navigate <https://help.ubuntu.com/community/UsingTheTerminal>`_ to the directory in which the ``.sh`` file is located and enter ``./file.sh``, where ``file.sh`` is to be replaced with the full name of the Anaconda download. This starts the installation process; from there, simply follow the prompts.
+ `In a terminal, navigate <https://help.ubuntu.com/community/UsingTheTerminal>`_ to the location in which you'd like to install Anaconda. You can download the file using the terminal command ``curl -O URL_OF_FILE``, where ``URL_OF_FILE`` is to be replaced by the URL of the Anaconda installer ``.sh`` file found at the Anaconda Distribution website. When the download is complete, enter ``chmod +x file.sh``, where ``file.sh`` is to be replaced with the full name of the downloaded file. This ensures that the file's permissions allow it to be executed. Next, enter ``./file.sh``, where ``file.sh`` is again to be replaced with the full name of the downloaded file. This starts the installation process. From there, simply follow the prompts, including the "Next Steps" at the end after the installer finishes.
```
Comment (endolith) [inline doc/source/dev/contributor/quickstart_ubuntu.rst]: You can also just run `bash Anaconda3-2019.03-Linux-x86_64.sh` instead of making it executable?  That's what it says on https://docs.anaconda.com/anaconda/install/linux/

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 2 above)
Comment (endolith) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: > A little unfortunate--I would have thought / hoped there was a way to install the older runtime libraries needed through conda-forge or something like that.

Agreed, though if it's only modifying a file within the development environment, that's not as bad.  The original instructions said to modify a file in the base environment.

File: doc/source/dev/contributor/quickstart_ubuntu.rst [hunk 4]
```diff
@@ -53,11 +56,13 @@ Building SciPy
...
-#. In the terminal, update all of SciPy's build dependencies: ``conda update setuptools wheel cython numpy matplotlib pytest pybind11``
+#. In the terminal, ensure that all of SciPy's build dependencies are up to date: ``conda install pybind11``, then ``conda update cython numpy matplotlib pytest pybind11``
 
-#. (Optional) Check your present working directory by entering ``pwd`` at the terminal. You should be in the root ``/scipy`` directory, not in a directory ending ``/scipy/scipy``.
+ .. note::
 
-#. Rename the file ``anaconda3/lib/libgfortran.so`` to ``anaconda3/lib/libgfortran.so_backup``. This file provides an incorrect Fortran complier; renaming it forces the system to find the Fortran compiler included with ``gcc`` instead.
+	If ``conda`` is not a recognized command, please see "Should I add Anaconda to the macOS or Linux PATH?" in the `Anaconda FAQ`_.
+
+#. (Optional) Check your present working directory by entering ``pwd`` at the terminal. You should be in the root ``/scipy`` directory, not in a directory ending ``/scipy/scipy``.
```
Comment (endolith) [inline doc/source/dev/contributor/quickstart_ubuntu.rst]:
```suggestion
#. (Optional) Check your present working directory by entering ``pwd`` at the terminal. You should be in the root ``.../scipy`` directory, not in a directory ending ``.../scipy/scipy``.
```

**Review by mdhaber — COMMENTED** [review]
@endolith yes, I've been thinking about creating some sort of image and planned to work on it more in early August.

**Comment by endolith** [comment]
So I'll move discussion to this issue instead of hijacking #8325.

@andyfaff Ok so [the Docker method](https://github.com/scipy/scipy/issues/8325#issuecomment-514116900) works well (behaves like a lightweight VM), and I can [share my git repo folder](https://github.com/scipy/scipy/issues/8325#issuecomment-514404030) with it, and it builds.

Any advice on how to best interact with it?  Seems like installing spyder etc into the container is not the way containers are meant to be used?  (They're generally meant for just running one process?)  Could run a spyder kernel inside the container and connect to it from outside?  But then editing the file isn't smooth: https://github.com/spyder-ide/spyder/issues/3065#issuecomment-511269829

Also, does it make sense for the `python setup.py build` `scipy pip install .` etc. to be moved into the Dockerfile?

**Comment by andyfaff** [comment]
> Ok so the Docker method works well (behaves like a lightweight VM), and I can share my git repo folder with it, and it builds.

Great.
The advantage to a Docker approach to get people started quickly is that once prospective developers have got Docker installed and running, then the rest of the steps are the same, whatever the host OS. The container provides all the necessary machinery, such as compilers, etc. Unfortunately any build artefacts are then tied to the container OS.

In comparison, the sets of steps to build and develop in each OS are obviously more diverse, and harder to keep up to date. However, it's vital that people do attempt builds in their own OS because it's the only way that they can get dev versions of scipy installed on their own computer, and it keeps the build ecosystem healthy.

> Any advice on how to best interact with it?

TLDR: it depends on how much interaction you need with the changes you're making. Very simple editing could be done from vi/emacs from within the container. However, my suggestion would be to use PyCharm (or whatever IDE you prefer) from the Host. If you mount the directory any files you change in the IDE will be visible in the container. When you've made enough changes you can then run the tests from the docker container, you don't need to run the whole test suite, running individual tests is possible with the `runtests.py` script.  It might be possible to install an IDE within the container then use X forwarding or VNC to see the GUI, but that definitely increases the level of difficulty. You could always start a Python session to explore the changes you make.

> does it make sense for the python setup.py build scipy pip install . etc. to be moved into the Dockerfile?

No.

- it takes a fair amount of time to compile scipy and it's better to build the docker container quickly
- the git repository from the host OS is only available in the container when you run the container with a mounted volume. I'm not sure if you can use a local volume whilst building the container. You could clone your remote github branch into the container, but then it's not working from the repository on your computer.
- there is a lot of churn in the scipy repository, you'd have to rebuild the container frequently
- I'm not sure if the docker build process notices changes to a scipy repository from the last time the build was run.
- if we want to make a Dockerfile available from a scipy project location for a quickstart purpose then we'd want the users to be mounting their own local clone of scipy. This is why I suggested removing the git clone step in the Dockerfile I provided.

**Comment by andyfaff** [comment]
@mdhaber I've just looked through the Ubuntu quickstart instructions. I'd say it's probably going to be a lot simpler to use a Docker container for a development quickstart on Linux.

**Comment by endolith** [comment]
@andyfaff 

> However, my suggestion would be to use PyCharm (or whatever IDE you prefer) from the Host. If you mount the directory any files you change in the IDE will be visible in the container. When you've made enough changes you can then run the tests from the docker container, you don't need to run the whole test suite, running individual tests is possible with the `runtests.py` script. 

Right, I have that running, but I mean when I'm making changes to the code and want to try them out in practical situations, etc.  (For example, I want to run timing comparisons of different implementations of the same algorithm and make graphs of it: https://github.com/scipy/scipy/issues/10427)

> It might be possible to install an IDE within the container then use X forwarding or VNC to see the GUI, but that definitely increases the level of difficulty. You could always start a Python session to explore the changes you make.

Hmm.  Sounds like running it as a spyder or jupyter server and running chunks of code at a time is the easiest.

>     * it takes a fair amount of time to compile scipy and it's better to build the docker container quickly

Would you ever want to do one and not the other though?

>     * the git repository from the host OS is only available in the container when you run the container with a mounted volume. I'm not sure if you can use a local volume whilst building the container. 

Oh, that makes sense.

> on Ubuntu, a shorter list of instructions (from a virgin install would be):

That would replace the system's scipy with scipy-dev, though, which could break things?  

I guess instead of keeping scipy-dev isolated inside a conda environment, I could instead keep the conda-downloaded scipy build isolated inside a conda environment and activate it when needed?

**Comment by mdhaber** [comment]
> I don't think any of the anaconda or homebrew steps are required.

No, the Anaconda steps aren't required. I find them helpful before trying to use conda, though.

And the Linuxbrew steps shouldn't be required, but I've found the Linuxbrew compilers to build SciPy more reliably than `apt` installed ones.

But yes, it would be great to _also_ have quickstart instructions that use pip instead of conda, and in any case it would be great if we can figure out steps to _reliably_ (on anyone's computer, without running into gh-8325 issues) build SciPy without Linuxbrew.

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 3 above)
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: Good idea.

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 4 above)
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: Made this change. Thanks.

**Comment by mdhaber** [comment]
@peterbell10  haha I tried `git rebase --onto master` and ended up with this! (I didn't intend to change 101 files....) Thoughts on recovery? I do have a backup of my branch from before the rebase, if that helps.

**Comment by peterbell10** [comment]
I just did `git rebase --onto upstream/master` on this branch locally and after resolving a few conflicts,  I get the following log from `git log upstream/master..quickstart_update`
```
commit b073020d8e2255db3dccdcc1c88f6563d45542e7
Author: Matt Haberland <mdhaber@mit.edu>
Date:   Tue Sep 3 13:54:59 2019 -0700

    DOC: various corrections to quickstart guides

commit 971e4016665729ae76449ce8f2344d3adf9a5a6a
Author: Matt Haberland <mdhaber@mit.edu>
Date:   Tue Sep 3 11:25:54 2019 -0700

    DOC: fixed a whitespace error in Ubuntu quickstart

commit 281788740c39c99b170ccdfcade11e7a0c06b341
Author: Matt Haberland <mhaberla@calpoly.edu>
Date:   Sun Aug 18 12:55:28 2019 -0700

    DOC: suggest Docker instructions for Windows users

commit 1eb6b9258b1a61cb5399a9c20597f3c190f931f2
Author: Matt Haberland <mhaberla@calpoly.edu>
Date:   Sun Aug 18 12:51:52 2019 -0700

    DOC: revised instructions according to reviewer comments

commit 86ce0ee7b2c036cfad5c0fcd14fca9d27e844059
Author: Matt Haberland <mhaberla@calpoly.edu>
Date:   Sun Jul 14 19:09:18 2019 -0500

    DOC: removed old build instructions
    
    The old build instructions are out of date. The information they
    contain might at some point be reorganized to help users install
    without Anaconda, but for now they should be removed because they
    don't work reliably without modification.

commit f61d82854539e0feff08440e034772e5109eb690
Author: Matt Haberland <mhaberla@calpoly.edu>
Date:   Sun Jul 14 18:13:19 2019 -0500

    DOC: revised Ubuntu quickstart guide for use from Windows
    
    Windows Subsystem for Linux and Amazon EC2 instances make it relatively
    easy to create and use a SciPy development environment from a Windows
    machine.

commit 5a759605c6b1e250aacac3773b95b7e7ca91de9b
Author: Matt Haberland <mhaberla@calpoly.edu>
Date:   Sun Jul 14 18:13:19 2019 -0500

    DOC: revised Ubuntu quickstart

```

>I was thinking that next time instead of `git rebase --onto master`, maybe `git reset --hard master` (locally) and then pull the commits back from the remote?

That would just create more merge commits.

File: doc/source/dev/contributor/quickstart_ubuntu.rst [hunk 5]
```diff
@@ -76,17 +83,17 @@ Currently we have *two* versions of SciPy: the latest release as installed by An
...
-
 #. Enter ``conda install cython numpy pytest spyder pybind11``. |br| Note that we're only installing SciPy's build dependencies (and Spyder so we can use the IDE), but not SciPy itself.
 
-#. Enter ``conda develop /scipy``, where ``scipy`` is to be replaced with the full path of the SciPy root directory. |br| This instructs ``conda`` to add the root SciPy directory to the |PYTHONPATH|_ environment variable whenever our ``scipydev`` virtual environment is activated. That way, when we ``import`` SciPy code, the code is imported from our development version of SciPy.
+#. Rename the file ``anaconda3/envs/scipydev/lib/libgfortran.so`` to ``anaconda3/envs/scipydev/lib/libgfortran.so_backup``, where ``anaconda3`` is to be replaced with the full path of your Anaconda installation. *Note: this ``libgfortran.so`` file is in a different location from the one we renamed already; it was created when we installed packages in the new virtual environment. This file needs to be renamed whenever setting up a new virtual environment for SciPy development.*
```
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst]: This step needed to be after installing build dependencies.

**Comment by rgommers** [comment]
Agree with @peterbell10's rebase advice. The `--onto` is optional, I normally type it as `git rebase upstream/master`. Pulling in random commits happens a lot, and it's hard to tell what people are doing wrong. Perhaps the original base branch wasn't master when the feature branch got created.

**Comment by mdhaber** [comment]
Well, I'll give it another shot next time something like this happens. For this one, a hard reset and pull from the backup did the trick. Thanks @peterbell10.

**Comment by mdhaber** [comment]
From my perspective, this is ready for squash merge if the documents render properly. The changes are mostly objective corrections, so while what was already in the guide is debatable, I think this is a needed update.

**Comment by mdhaber** [comment]
Oh, the joy of ReST: 30 minutes waiting for builds to verify that an invisible character turns:
Screen Shot 2019-09-03 at 10 14 42 PM
into:
Screen Shot 2019-09-03 at 10 33 18 PM

**Comment by mdhaber** [comment]
@tylerjereddy Would you take a look at these changes? It fixes an important mistake in the instructions for renaming the `libgfortran.so` file.

File: doc/source/dev/contributor/quickstart_mac.rst [hunk 6]
```diff
@@ -51,7 +51,7 @@ Building SciPy
 
 #. Use Homebrew to install ``gcc`` by entering the command ``brew install gcc``.
 
-#. In the terminal, update/upgrade all of SciPy's build dependencies: ``conda update cython numpy pytest pybind11``
+#. In the terminal, ensure that all of SciPy's build dependencies are up to date: ``conda install pybind11``, then ``conda update cython numpy pytest pybind11``.
```
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_mac.rst]: `pybind` is not distributed with Anaconda, so the user will not yet have it.

File: doc/source/dev/contributor/quickstart_mac.rst [hunk 7]
```diff
@@ -80,7 +80,7 @@ Currently we have *two* versions of SciPy: the latest release as installed by An
...
 
-#. Enter ``conda develop /scipy``, where ``scipy`` is to be replaced with the full path of the SciPy root directory. |br| This instructs ``conda`` to add the root SciPy directory to the |PYTHONPATH|_ environment variable whenever our ``scipydev`` virtual environment is activated. That way, when we ``import`` SciPy code, the code is imported from our development version of SciPy.  *Note: this step differs from the process shown in* `Anaconda SciPy Dev: Part II (macOS)`_ *, which involves creating special folders and files deep within the Anaconda directory. You can ignore that part of the video from 0:38 to 1:38; this is much simpler!*
+#. Enter ``conda develop /scipy``, where ``scipy`` is to be replaced with the full path of the SciPy root directory. |br| This will allow us to ``import`` the development version of SciPy in Python regardless of Python's working directory. *Note: this step replace the steps shown in* `Anaconda SciPy Dev: Part II (macOS)`_ *that modify the ``PYTHONPATH`` environment variable when the ``scipydev`` virtual environment is activated. You can ignore that part of the video from 0:38 to 1:38; this is much simpler!*
```
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_mac.rst]: Removed the details about `PYTHONPATH` environment variable, because I don't think that's how `conda develop` actually works.

File: doc/source/dev/contributor/quickstart_ubuntu.rst [hunk 8]
```diff
@@ -2,9 +2,9 @@
 
 .. _quickstart-ubuntu:
 
-=================================================
-Development Environment Quickstart Guide (Ubuntu)
-=================================================
+=======================================================
+Development Environment Quickstart Guide (Ubuntu 16.04)
+=======================================================
```
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst]: Might work with other versions, but this procedure has not failed for me in 16.04 on several machines (native linux, Amazon EC2 instance, within Docker container on Windows, WSL on Windows).

File: doc/source/dev/contributor/quickstart_ubuntu.rst [hunk 9]
```diff
@@ -53,11 +58,13 @@ Building SciPy
 
 #. Use Homebrew to install ``gcc`` by entering the command ``brew install gcc``.
 
-#. In the terminal, update all of SciPy's build dependencies: ``conda update cython numpy pytest pybind11``
+#. In the terminal, ensure that all of SciPy's build dependencies are up to date: ``conda install pybind11``, then ``conda update cython numpy pytest pybind11``.
+
+   .. note::
 
-#. (Optional) Check your present working directory by entering ``pwd`` at the terminal. You should be in the root ``/scipy`` directory, not in a directory ending ``/scipy/scipy``.
+      If ``conda`` is not a recognized command, try restarting your terminal. If it is still not recognized, please see "Should I add Anaconda to the macOS or Linux PATH?" in the `Anaconda FAQ`_.
```
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst]: conda's initialization procedure doesn't take effect until terminal has been restarted.

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 1 above)
Comment (Kai-Striega) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: @tylerjereddy regarding (2) I've been using visual studio code with the [Remote - WSL](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl) plugin and it's worked well for me so far.

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: Ooh that sounds like exactly what I want Spyder to do. I will try this in the meantime.

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: Wow, this is great. Finally, a relatively simple and complete way of doing SciPy development from Windows.

File: doc/source/dev/contributor/quickstart_mac.rst (hunk 6 above)
Comment (tylerjereddy) [reply doc/source/dev/contributor/quickstart_mac.rst]: ok

File: doc/source/dev/contributor/quickstart_mac.rst (hunk 7 above)
Comment (tylerjereddy) [reply doc/source/dev/contributor/quickstart_mac.rst]: "this step replace the steps shown" -> should that be ... replaces ...?

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 5 above)
Comment (tylerjereddy) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: Just to clarify, our instructions involve moving *two* `libgfortran.so` files now?

**Review by tylerjereddy — COMMENTED** [review]
There's quite a lot of discussion in the PR, but looking at the diff now it seems to be converging on a relatively straightforward update of the docs?

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 2 above)
Comment (tylerjereddy) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: > Agreed, though if it's only modifying a file within the development environment, that's not as bad. The original instructions said to modify a file in the base environment.

Isn't it both now?

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 1 above)
Comment (tylerjereddy) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: This is cool stuff--I'll mark as resolved though so we don't get too side-tracked on this specific PR.

File: doc/source/dev/contributor/quickstart_mac.rst (hunk 7 above)
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_mac.rst]: Yes; typo.

Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_mac.rst]:
```suggestion
#. Enter ``conda develop /scipy``, where ``scipy`` is to be replaced with the full path of the SciPy root directory. |br| This will allow us to ``import`` the development version of SciPy in Python regardless of Python's working directory. *Note: this step replaces the steps shown in* `Anaconda SciPy Dev: Part II (macOS)`_ *that modify the ``PYTHONPATH`` environment variable when the ``scipydev`` virtual environment is activated. You can ignore that part of the video from 0:38 to 1:38; this is much simpler!*
```

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 2 above)
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: > Isn't it both now?

Yes. According to the instructions, we build SciPy first in the base environment to keep the number of instructions from step 0 to an initial run of the test suite just a bit smaller. We need to rename the file in the base environment to do that. Then we create the virtual environment and rename the file there, too.

Technically, we could skip all the steps that are performed in the base environment, and only build SciPy in the development environment, and then we'd only need to modify the file in the development environment. 

Besides (slightly) reducing the number of instructions from step 0 to running the test suite, another slight advantage of the current approach is that it introduces virtual environments separately from the main process of building SciPy, hopefully giving the beginner a better sense of what the individual steps are really doing. That was part of the original intent, anyway. In a future revision, we could reconsider this approach, but it should probably also be changed for the macOS instructions, and then the tutorial would diverge a bit more from the video.

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 5 above)
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: Yes

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 2 above)
Comment (rgommers) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: > Yes. According to the instructions, we first build SciPy in the base environment. We need to rename the file in the base environment to do that

This is very unhealthy, please don't recommend that. Having to do it in an env other than `base` is bad enough, but don't mess with `base`.

> Technically, we could skip all the steps that are performed in the base environment, and only build SciPy in the development environment, and then we'd only need to modify the file in the development environment.

+1

**Comment by larsoner** [comment]
@mdhaber it looks like there is a conflict here now. Since this PR is so short, I propose that once you fix that I take a final look and (so long as it seems uncontroversial) I'll merge

File: doc/source/dev/contributor/quickstart_mac.rst [hunk 10]
```diff
@@ -42,67 +42,68 @@ Building SciPy
 *Consider following along with the companion video* `Anaconda SciPy Dev: Part I (macOS)`_
 
 #. Download, install, and test the latest release of the `Anaconda Distribution of Python`_.
-In addition to the latest version of Python 3, the Anaconda distribution includes
-dozens of the most popular Python packages for scientific computing, the Spyder
-integrated development environment (IDE), the ``conda`` package manager, and tools
-for managing virtual environments.
+   In addition to the latest version of Python 3, the Anaconda distribution includes
```
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_mac.rst]: These spaces are required to make this render as a numbered list.

File: doc/source/dev/contributor/quickstart_ubuntu.rst [hunk 11]
```diff
@@ -2,9 +2,9 @@
 
 .. _quickstart-ubuntu:
 
-=================================================
-Development Environment Quickstart Guide (Ubuntu)
-=================================================
+=======================================================
+Development environment quickstart guide (Ubuntu 16.04)
```
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst]: Apparently section headings are supposed to be lower case? This is very inconsistent, but the other quickstarts were changed in gh-10823 so I guess this should follow suit.

**Comment by mdhaber** [comment]
@larsoner I tried to stick to just fixing merge conflicts, but I noticed that the numbered list in the macOS quickstart was [not rendering](http://scipy.github.io/devdocs/dev/contributor/quickstart_mac.html#quickstart-mac) so I fixed that, too.

I have not changed the instructions to avoid having to alter the conda base environment `libgfortran.so` file. To keep the diff readable, can I do that in a second PR?

File: doc/source/dev/contributor/quickstart_ubuntu.rst (hunk 2 above)
Comment (larsoner) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: > I have not changed the instructions to avoid having to alter the conda base environment libgfortran.so file. To keep the diff readable, can I do that in a second PR?

Another PR to get the renaming stuff correct is fine, but I would just get rid of the sections about renaming entirely here in that case.

But really I'm not too worried about the diff becoming worse, it's not so bad currently. Clicking the "hide whitespace changes" makes it even more manageable.

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: OK. I'll try to change it tonight.

**Comment by rgommers** [comment]
The instructions in current master are already wrong, so I'm just going to merge this. But we should really address two issues in a follow-up:

1. no building in base, building twice is very odd
2. no messing with `libgfortran`. the problem is probably with Homebrew, using XCode compilers or conda compilers things work fine.

**Comment by rgommers** [comment]
Merged, thanks @mdhaber, all!

**Comment by mdhaber** [comment]
I will change the tutorials to not build in base.

I didn't want to use Homebrew here. I know you didn't recommend it, @rgommers, but I had even more trouble with other compilers. I can try again sometime with other compilers, but it's important to find a procedure that will work for everyone, and this was the most robust I've found.

**Comment by rgommers** [comment]
Actually, Homebrew is fine. I double checked, I have `gfortran` 8.3 from Homebrew, and also 
```
$ ls ~/anaconda3/lib/libgf*
/Users/rgommers/anaconda3/lib/libgfortran.3.dylib
/Users/rgommers/anaconda3/lib/libgfortran.dylib
```
those do not conflict. 

> it's important to find a procedure that will work for everyone, and this was the most robust I've found.

yeah that's why I'm worried, this is definitely not robust. If you manually delete `libgfortran` you've just broken every installed package that relies on it.
